File: Week 03/id_493/127.单词接龙.js
/*
 * @lc app=leetcode.cn id=127 lang=javascript
 *
 * [127] 单词接龙
 */

// @lc code=start
/**
 * @param {string} beginWord
 * @param {string} endWord
 * @param {string[]} wordList
 * @return {number}
 */
var ladderLength = function(beginWord, endWord, wordList) {

    let dict = {};
    for (let word of wordList) {
        for (let i = 0; i < word.length; i++) {
            let key = word.substring(0, i) + "*" + word.substring(i + 1);
            dict[key] = dict[key] || new Set();
            dict[key].add(word);
        }
    }    
    let visit = [];
    let findSet = new Set([beginWord]);
    let len = 1;
    while (findSet.size > 0) {
        let childList = new Set();
        len++;
        for (let word of findSet) {
            for (let i = 0; i < word.length; i++) {
                let key = word.substring(0, i) + "*" + word.substring(i + 1);
                if (dict.hasOwnProperty(key)) {
                    let ff = dict[key];
                    if (ff.has(endWord)) {
                        return len;
                    }
                    for (let cc of ff) {
                        if (visit.indexOf(cc) == -1) {
                            childList.add(cc);
                            visit.push(cc)
                        }
                        
                    }
                    
                }
            }
        }
        findSet = childList;
    }
    return 0;
};

// @lc code=end

File: Week 06/id_158/127.单词接龙.cs
/*
 * @lc app=leetcode.cn id=127 lang=csharp
 *
 * [127] 单词接龙
 */

// @lc code=start
using System.Collections.Generic;

public class Solution
{
    public int LadderLength(string beginWord, string endWord, IList<string> wordList)
    {
        if (!wordList.Contains(endWord))
            return 0;
        int n = beginWord.Length;
        Dictionary<string, IList<string>> allCommons = new Dictionary<string, IList<string>>();
        foreach (var word in wordList)
        {
            for (int i = 0; i < n; i++)
            {
                string common = word.Substring(0, i) + "*" + word.Substring(i + 1);
                if (!allCommons.ContainsKey(common))
                    allCommons.Add(common, new List<string>());
                allCommons[common].Add(word);
            }
        }
        IList<string> begin = new List<string>();
        IList<string> end = new List<string>();
        begin.Add(beginWord);
        end.Add(endWord);

        IList<string> visited = new List<string>();
        int len = 1;
        while (begin.Count > 0 && end.Count > 0)
        {
            if (begin.Count > end.Count)
            {
                IList<string> tmp = begin;
                begin = end;
                end = tmp;
            }
            IList<string> neighbor = new List<string>();
            foreach (string cur in begin)
            {
                for (int i = 0; i < n; i++)
                {
                    string tmp = cur.Substring(0, i) + "*" + cur.Substring(i + 1);
                    if (!allCommons.ContainsKey(tmp))
                        continue;
                    foreach (string str in allCommons[tmp])
                    {
                        if (end.Contains(str))
                            return len + 1;
                        if (!visited.Contains(str))
                        {
                            visited.Add(str);
                            neighbor.Add(str);
                        }
                    }
                }
            }
            begin = neighbor;
            len++;
        }
        return 0;
    }
}
// @lc code=end

